apps/perl/5.18.2/man/man3/version.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "version 3"
.TH version 3 "2013-08-16" "perl v5.18.2" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
version \- Perl extension for Version Objects
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&  # Parsing version strings (decimal or dotted\-decimal)
\&
\&  use version 0.77; # get latest bug\-fixes and API
\&  $ver = version\->parse($string)
\&
\&  # Declaring a dotted\-decimal $VERSION (keep on one line!)
\&
\&  use version; our $VERSION = version\->declare("v1.2.3"); # formal
\&  use version; our $VERSION = qv("v1.2.3");               # shorthand
\&  use version; our $VERSION = qv("v1.2_3");               # alpha
\&
\&  # Declaring an old\-style decimal $VERSION (use quotes!)
\&
\&  our $VERSION = "1.0203";                                # recommended
\&  use version; our $VERSION = version\->parse("1.0203");   # formal
\&  use version; our $VERSION = version\->parse("1.02_03");  # alpha
\&
\&  # Comparing mixed version styles (decimals, dotted\-decimals, objects)
\&
\&  if ( version\->parse($v1) == version\->parse($v2) ) {
\&    # do stuff
\&  }
\&
\&  # Sorting mixed version styles
\&
\&  @ordered = sort { version\->parse($a) <=> version\->parse($b) } @list;
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Version objects were added to Perl in 5.10.  This module implements version
objects for older version of Perl and provides the version object \s-1API\s0 for all
versions of Perl.  All previous releases before 0.74 are deprecated and should
not be used due to incompatible \s-1API\s0 changes.  Version 0.77 introduces the new
\&'parse' and 'declare' methods to standardize usage.  You are strongly urged to
set 0.77 as a minimum in your code, e.g.
.PP
.Vb 1
\&  use version 0.77; # even for Perl v.5.10.0
.Ve
.SH "TYPES OF VERSION OBJECTS"
.IX Header "TYPES OF VERSION OBJECTS"
There are two different types of version objects, corresponding to the two
different styles of versions in use:
.IP "Decimal Versions" 2
.IX Item "Decimal Versions"
The classic floating-point number \f(CW$VERSION\fR.  The advantage to this style is
that you don't need to do anything special, just type a number into your
source file.  Quoting is recommended, as it ensures that trailing zeroes
(\*(L"1.50\*(R") are preserved in any warnings or other output.
.IP "Dotted Decimal Versions" 2
.IX Item "Dotted Decimal Versions"
The more modern form of version assignment, with 3 (or potentially more)
integers separated by decimal points (e.g. v1.2.3).  This is the form that
Perl itself has used since 5.6.0 was released.  The leading 'v' is now
strongly recommended for clarity, and will throw a warning in a future
release if omitted.  A leading 'v' character is required to pass the
\&\*(L"\fIis_strict()\fR\*(R" test.
.SH "DECLARING VERSIONS"
.IX Header "DECLARING VERSIONS"
If you have a module that uses a decimal \f(CW$VERSION\fR (floating point), and you
do not intend to ever change that, this module is not for you.  There is
nothing that version.pm gains you over a simple \f(CW$VERSION\fR assignment:
.PP
.Vb 1
\&  our $VERSION = "1.02";
.Ve
.PP
Since Perl v5.10.0 includes the version.pm comparison logic anyways,
you don't need to do anything at all.
.SS "How to convert a module from decimal to dotted-decimal"
.IX Subsection "How to convert a module from decimal to dotted-decimal"
If you have used a decimal \f(CW$VERSION\fR in the past and wish to switch to a
dotted-decimal \f(CW$VERSION\fR, then you need to make a one-time conversion to
the new format.
.PP
\&\fBImportant Note\fR: you must ensure that your new \f(CW$VERSION\fR is numerically
greater than your current decimal \f(CW$VERSION\fR; this is not always obvious. First,
convert your old decimal version (e.g. 1.02) to a normalized dotted-decimal
form:
.PP
.Vb 2
\&  $ perl \-Mversion \-e \*(Aqprint version\->parse("1.02")\->normal\*(Aq
\&  v1.20.0
.Ve
.PP
Then increment any of the dotted-decimal components (v1.20.1 or v1.21.0).
.ie n .SS "How to ""declare()"" a dotted-decimal version"
.el .SS "How to \f(CWdeclare()\fP a dotted-decimal version"
.IX Subsection "How to declare() a dotted-decimal version"
.Vb 1
\&  use version; our $VERSION = version\->declare("v1.2.3");
.Ve
.PP
The \f(CW\*(C`declare()\*(C'\fR method always creates dotted-decimal version objects.  When
used in a module, you \fBmust\fR put it on the same line as \*(L"use version\*(R" to
ensure that \f(CW$VERSION\fR is read correctly by \s-1PAUSE\s0 and installer tools.  You
should also add 'version' to the 'configure_requires' section of your
module metadata file.  See instructions in ExtUtils::MakeMaker or
Module::Build for details.
.PP
\&\fBImportant Note\fR: Even if you pass in what looks like a decimal number
(\*(L"1.2\*(R"), a dotted-decimal will be created (\*(L"v1.200.0\*(R"). To avoid confusion
or unintentional errors on older Perls, follow these guidelines:
.IP "\(bu" 2
Always use a dotted-decimal with (at least) three components
.IP "\(bu" 2
Always use a leading-v
.IP "\(bu" 2
Always quote the version
.PP
If you really insist on using version.pm with an ordinary decimal version,
use \f(CW\*(C`parse()\*(C'\fR instead of declare.  See the \*(L"\s-1PARSING AND COMPARING VERSIONS\*(R"\s0
for details.
.PP
See also version::Internals for more on version number conversion,
quoting, calculated version numbers and declaring developer or \*(L"alpha\*(R" version
numbers.
.SH "PARSING AND COMPARING VERSIONS"
.IX Header "PARSING AND COMPARING VERSIONS"

lib/Agua/Ops/Version.pm
package Agua::Ops::Version;
use Moose::Role;
use Method::Signatures::Simple;
use JSON;

=head3	PACKAGE		Version
	
PURPOSE

    1. UPDATE THE VERSION NUMBER, PRE-RELEASE LABEL OR BUILD IN
    
        THE VERSION FILE ACCORDING TO THE PROVIDED UPDATE TYPE
        
    2. PRINT THE USER-SUPPLIED DESCRIPTION TO THE VERSION FILE 
    
    3. ADD THE DATETIME TO THE VERSION FILE
    
    4. ADD A TAG USING THE WHOLE VERSION AND DESCRIPTION

INPUT

    1. BASE DIR OF git REPOSITORY
    
    2. UPDATE TYPE: major, minor, patch OR build

    3. (OPTIONAL) RELEASE NAME

    3. DESCRIPTION OF VERSION
    
OUTPUT

    1. UPDATED VERSION FILE IN BASE DIR OF git REPOSITORY
    
    2. GIT COMMITTED TAG CONTAINING WHOLE VERSION AND DESCRIPTION

NOTES

    ONLY USE THIS APPLICATION IF YOU WANT TO CREATE A NEW TAG TO RELEASE
    
    BUG FIXES (PATCH), NEW FEATURES (MINOR VERSION) OR API-RELATED CHANGES
    
    {MAJOR VERSION), OR TO MARK A PRE-RELEASE VERSION (RELEASE) OR A SMALL 
    
    INCREMENTAL IMPROVEMENT ON ANY OF THE ABOVE (BUILD).
    
    IF NONE OF THE ABOVE APPLY, DO NOT USE THIS APPLICATION. INSTEAD, TRACK
    
    CHANGES WITH 'git log' USING THE COMMIT ITERATION COUNTER, BUILD ID AND COMMENTS.

LICENCE

This code is released under the MIT license, a copy of which should
be provided with the code.

=cut

use strict;
use warnings;

#### USE LIB
use FindBin qw($Bin);
use lib "$Bin/../../lib";

#### EXTERNAL MODULES
use Data::Dumper;

#### INTERNAL MODULES
use Agua::DBaseFactory;
use Agua::Ops;

#### Strings
has 'versionformat'	=> ( isa => 'Str', is  => 'rw', default	=>	'semver'	);

#####////}}}}

#### SORT VERSIONS
method sortVersions ($versions) {
	$self->logCaller("");
	$self->logNote("BEFORE versions: @$versions");

	sub splitStringNumber () {
		my $string		=	shift;
		#print "splitStringNumber    string: $string\n";
		my $stringObject = {};
		($stringObject->{string}) = $string =~ /^([^\d^\.]+)/;
		($stringObject->{number}) = $string =~ /(\d+)$/;

		return $stringObject;
	}
	
	sub compareStringNumber () {
		my $a		=	shift;
		my $b		=	shift;
		#print "compareStringNumber    a: $a\n";
		#print "compareStringNumber    b: $b\n";

		if ( $a !~ /(\d+)$/ and $b !~ /(\d+)$/ ) {
			my $compare = lc($a) cmp lc($b);
			#print "compareStringNumber    compare: $compare\n";
			return $compare;
		}
		else {
			my $aObject = &splitStringNumber($a);
			my $bObject = &splitStringNumber($b);
			#print "aObject: $aObject\n";
			#print "bObject: $bObject\n";

			if ( $aObject->{string} ne $bObject->{string} ) {
				my $compare = lc($a) cmp lc($b);
				return $compare;
			}
			else {
				#print "comparing numbers\n";	
				if ( not defined $aObject->{number} and not defined $bObject->{number} ) {
					return 0;
				}
				elsif ( $aObject->{number} and not defined $bObject->{number} ) {
					return 1;
				}
				elsif ( $bObject->{number} and not defined $aObject->{number} ) {
					return -1;
				}
				elsif ( $aObject->{number} > $bObject->{number} ) {
					#print "$aObject->{number} is larger than $bObject->{number}\n";
					return 1;
				}
				elsif ( $bObject->{number} > $aObject->{number} ) {
					#print "$bObject->{number} is larger than $aObject->{number}\n";
					return -1;
				}
				else {
					return 0;
				}
			}
		}
	}

	sub parseVersion {
		my $version 	=	shift;
	
		$version =~
			/^(\d+)\.(\d+)\.(\d+)(-alpha[\.\d]*|-beta[\.\d]*|-rc[\.\d]*)?(\+build[\.\d]*|\+[\.\d]*)?/;
	
		my $object = {};
		$object->{major} 	= $1;
		$object->{minor} 	= $2;
		$object->{patch} 	= $3;
		$object->{release} 	= $4;
		$object->{build}  	= $5;
		$object->{build} =~ s/\.+//g if defined $object->{build};
		$object->{release} =~ s/\.+//g if defined $object->{release};
		$object->{build} =~ s/^[\+\.]// if defined $object->{build};
		$object->{release} =~ s/^\-// if defined $object->{release};
		$object->{release} 	= '' if not defined $object->{release};
		$object->{build}  	= '' if not defined $object->{build};
	
		return $object;
	}

	sub versionSort {
		my $aVersion = &parseVersion($a);
		my $bVersion = &parseVersion($b);
	
		if ( $aVersion->{major} > $bVersion->{major} )	{ return 1 }
		elsif ( $bVersion->{major} > $aVersion->{major} ) { return -1 }
		if ( $aVersion->{minor} > $bVersion->{minor} )	{ return 1 }
		elsif ( $bVersion->{minor} > $aVersion->{minor} ) { return -1 }
		if ( $aVersion->{patch} > $bVersion->{patch} )	{ return 1 }
		elsif ( $bVersion->{patch} > $aVersion->{patch} ) { return -1 }
		if ( ! $aVersion->{release} and ! $bVersion->{release}
			and ! $aVersion->{build} and ! $bVersion->{build} )	{ return 0 }
		
		if ( $aVersion->{release} and ! $bVersion->{release} )	{ return -1 }
		if ( $bVersion->{release} and ! $aVersion->{release} )	{ return 1 }
		
		if ( $aVersion->{release} and $bVersion->{release} ) {
			my $compare = &compareStringNumber($aVersion->{release}, $bVersion->{release});
			#print "compareStringNumber(aVersion->release, bVersion->release): $compare\n";
			return $compare if $compare != 0;
			return 0 if ! $aVersion->{build} and ! $bVersion->{build};
		}
		
		if ( $aVersion->{build} and ! $bVersion->{build} )	{ return 1 }
		if ( $bVersion->{build} and ! $aVersion->{build} )	{ return -1 }
		if ( $aVersion->{build} and $bVersion->{build} ) {
			return &compareStringNumber($aVersion->{build}, $bVersion->{build});
		}
		
		return 0;
	}
	
	@$versions = sort versionSort @$versions if defined $versions and scalar(@$versions) > 1;
	$self->logNote("AFTER versions @$versions");
	
	return $versions;
}

method latestVersion ($owner, $repository, $privacy) {
	$self->logDebug("owner", $owner);
	$self->logDebug("repository", $repository);
	$self->logDebug("privacy", $privacy);
	
	my $tags 	=	$self->getRemoteTags($owner, $repository, $privacy);
	$self->logDebug("tags", $tags);

	my $versions = [];
	foreach my $tag ( @$tags ) {
		push @$versions, $tag->{name};
	}
	$versions = $self->sortVersions($versions);
	$self->logDebug("versions", $versions);

	#### SORT: FIRST TO LAST
	@$versions = reverse(@$versions);
	$self->logDebug("versions", $versions);

	return shift @$versions;
}

method higherVersion ($version1, $version2) {
#### BASIC NUMBER COMPARISON OR STRING/NUMBER COMPARISON

	$self->logDebug("PASSED version1", $version1);
	$self->logDebug("PASSED version2", $version2);
	
	if ( $version1 =~ /^\d+(\.\d+)?$/ and $version2 =~ /^\d+(\.\d+)?$/ ) {
		
		return 1 if $version1 > $version2;
		return 0;
	}
	
	my $comparison = &compareStringNumber($version1, $version2);

	return 1 if $comparison > 0;
	return 0;
}

method sameHigherVersion ($version1, $version2) {
#### COMPARISON OF X.X.XX.X VERSION DIGITS IN version1 AND version2
#### RETURNS 1 IF version2 IS SAME OR HIGHER

	$self->logDebug("PASSED version1", $version1);
	$self->logDebug("PASSED version2", $version2);
	my ($array1, $array2);
	@$array1 = split "\\.", $version1;
	$self->logDebug("array1", $array1);
	@$array2 = split "\\.", $version2;
	$self->logDebug("array2", $array2);
	
	for ( my $i = 0; $i < @$array1; $i++ ) {
		return 0 if not defined $$array2[$i];
		return 0 if $$array2[$i] eq "" and $$array1[$i] ne "";
		my ($value1) = $$array1[$i]	=~	/(\d+)/;
		my ($value2) = $$array2[$i]	=~	/(\d+)/;
		$self->logDebug("Comparing value1 $value1 to value2 $value2");
		
		return 0 if $value2 < $value1;
	}
	
	return 1;
}

method higherSemVer ($version1, $version2) {
	$self->logDebug("PASSED version1", $version1);
	$self->logDebug("PASSED version2", $version2);

	#### REMOVE ANYTHING AFTER +build.\d+
	$version1 =~ s/(build\.\d+)\..+$/$1/;
	$version2 =~ s/(build\.\d+)\..+$/$1/;
	#$self->logDebug("version1", $version1);
	#$self->logDebug("version2", $version2);
	
	return 0 if $version1 eq $version2;
	return 1 if defined $version1 and not defined $version2;
	return -1 if not defined $version1 and defined $version2;
	
	my $array = [ $version1, $version2 ];
	my @temp = @$array;
	my $sortedarray;
	if ( $#temp > 0 ) {
		$sortedarray = $self->sortVersions(\@temp);
	}
	elsif ( $#temp == 0 ) {
		$sortedarray = [$temp[0]];
	}
	$self->logDebug("array: @$array");
	$self->logDebug("sortedarray: @$sortedarray");
	
	if ( $self->arraysHaveSameOrder($array, $sortedarray) ) {
		$self->logDebug("returning -1");
		return -1;
	}
	else {
		$self->logDebug("returning 1");
		return 1;
	}
}

method arraysHaveSameOrder ($arrayA, $arrayB) {
	if ( scalar(@$arrayA) != scalar(@$arrayB) ) { return 0; }
	
	for (my $i = 0; $i < @$arrayA; $i++) {
		if ($$arrayA[$i] ne $$arrayB[$i]) { 
			return 0;
		}
	}
	return 1;
}


#### SET VERSION
method setVersion ($versionformat, $repodir, $versionfile, $branch, $version, $description) {
	$self->logDebug("");
	my ($result, $error);
	if ( $versionformat eq "semver" ) {
		($result, $error) = $self->setSemVer($repodir, $versionfile, $branch, $version, $description);
		$self->logDebug("result", $result);
		$self->logDebug("error", $error);
		return ($result, $error);
	}
	
	return $self->setNonSemVer($repodir, $versionfile, $branch, $version, $description);
}

method _setVersion($repodir, $versionfile, $branch, $version, $description) {
	$self->logCaller("");
	$self->logDebug("repodir", $repodir);
	$self->logDebug("versionfile", $versionfile);
	$self->logDebug("branch", $branch);
	$self->logDebug("version", $version);
	$self->logDebug("description", $description);

	#### WRITE VERSION FILE
	$self->printToFile($versionfile, $version);

	#### REMOVE LOCK IF PRESENT
	my $lockfile = "$repodir/.git/index.lock";
	`rm -fr $lockfile` if -f $lockfile;
	$self->logWarning(qq{Could not remove lockfile: $lockfile\nRun this command to commit: git commit -a -m "$description"}) if -f $lockfile;

	#### ADD CHANGED VERSION TO REPO
	$self->addToRepo();
	
	$self->logDebug("Returning ($version, undef)");
	return ($version, undef);
}

method getCurrentVersion ($repodir, $branch) {
	$self->logDebug("");

	#### GET CURRENT VERSION TAG
	$self->changeToRepo($repodir);
	my $command = "git describe --abbrev=0 --tag";
	$self->logDebug("command", $command);
	my ($version) = $self->runCommand($command);
	
	$self->logDebug("Returning current version", $version);
	return $version;
}

method setSemVer ($repodir, $versionfile, $branch, $version, $description) {
	$self->logDebug("version", $version);
	return if not $self->isSemVer($version);

	#### GET CURRENT VERSION
	my $currentversion = $self->getCurrentVersion($repodir, $branch);
	$self->logDebug("currentversion", $currentversion);
	
	my $error;
	($version, $error) = $self->validateSemVer($currentversion, $version);
	$self->logDebug("version", $version);
	$self->logDebug("error", $error);
	return (undef, $error) if not defined $version;

	#### SET VERSION
	return $self->_setVersion($repodir, $versionfile, $branch, $version, $description);
}

method setNonSemVer ($repodir, $versionfile, $branch, $version, $description) {
	#### GET CURRENT VERSION
	my $currentversion = $self->getCurrentVersion($repodir, $branch);
	#$self->logDebug("currentversion", $currentversion);	

	#### RETURN IF VERSIONS ARE THE SAME
	return if $version eq $currentversion;
	
	#### RETURN IF VERSION IS LOWER THAN THE CURRENT VERSION
	return if not $self->higherVersion($version, $currentversion);
	
	#### SET VERSION
	return $self->_setVersion($repodir, $versionfile, $branch, $version, $description);
}

method validateSemVer ($currentversion, $version) {
	if ( defined $currentversion and $currentversion) {
		#### REMOVE ANYTHING AFTER +build.\d+
		my $version1 = $version;
		my $version2 = $currentversion;
		$version1 =~ s/(build\.\d+)\..+$/$1/;
		$version2 =~ s/(build\.\d+)\..+$/$1/;
		$self->logDebug("version1", $version1);
		$self->logDebug("version2", $version2);

		#### RETURN IF VERSIONS ARE THE SAME
		return (undef, "Current version $currentversion is the same as specified version: $version") if $version1 eq $version2;
		
		#### RETURN IF VERSION IS LOWER THAN THE CURRENT VERSION
		$self->logDebug("self->higherSemVer($version1, $version2) ", $self->higherSemVer($version1, $version2));
		if ( $self->higherSemVer($version1, $version2) == 0 ) {
			return (undef, "Cannot set version because current version $currentversion is identical to specified version: $version");
		}
		elsif ( $self->higherSemVer($version1, $version2) < 0 ) {
			return (undef, "Cannot set version because current version $currentversion is higher than specified version: $version");
		}
	}

	return ($version, undef);	
}

method isSemVer ($version) {
	$self->logDebug("");
	
	#### PARSE OUT VERSIONS	
	my ($major, $minor, $patch, $release, $build) = $self->parseSemVer($version);
	#$self->logDebug("major", $major);
	#$self->logDebug("minor", $minor);
	#$self->logDebug("patch", $patch);
	#$self->logDebug("release", $release);
	#$self->logDebug("build", $build);

	return -1 if not defined $major or not $major;
	return -1 if not defined $minor or not $minor;
	return -1 if not defined $patch or not $patch;
	return -1 if not $release =~ /^(-alpha\.\d+|-beta\.\d+|-rc\.\d+)$/;
	return -1 if not $build =~ /^\+build\.\d+$/;	
	return 1;
}


#### INCREMENT VERSION
method incrementVersion ($versionformat, $versiontype, $repodir, $versionfile, $releasename, $description, $branch) {
	#### GET CURRENT VERSION
	my $currentversion = $self->getCurrentVersion($repodir, $branch);
	$self->logCritical("No current version found") and exit if not defined $currentversion or not $currentversion;

	#### INCREMENT VERSION	
	my $finalversion;
	if ( $versionformat eq "semver" ) {
		($finalversion) = $self->incrementSemVer($currentversion, $versiontype, $releasename);
	}
	else {
		$self->logDebug("Cannot increment version because versionformat not supported", $versionformat);
	}
	$self->logDebug("finalversion", $finalversion);

	#### MAKE CHANGES IF VERSION IS DEFINED
	if ( defined $finalversion ) {
		my ($result, $error) = $self->_setVersion($repodir, $versionfile, $branch, $finalversion, $description);
		$self->logDebug("result", $result);
		$self->logDebug("error", $error);
		return $result;
	}
	return;
}

method incrementSemVer ($currentversion, $versiontype, $releasename) {
#$self->logDebug("versiontype", $versiontype);
#$self->logDebug("releasename", $releasename);

	#### PARSE OUT VERSIONS	
	my ($major, $minor, $patch, $release, $build) = $self->parseSemVer($currentversion);
	$self->logDebug("major", $major);
	$self->logDebug("minor", $minor);
	$self->logDebug("patch", $patch);
	$self->logDebug("release", $release);
	$self->logDebug("build", $build);
	$self->logCritical("major version not defined in current version: $currentversion") and exit if not defined $major;
	$self->logCritical("minor version not defined in current version: $currentversion") and exit if not defined $minor;
	$self->logCritical("patch version not defined in current version: $currentversion") and exit if not defined $patch;

	#### SANITY CHECK
	if ( not defined $major ) {
		$self->logDebug("major version not defined");
		return;
	}

	if ( defined $releasename and $releasename ) {
		#### VERSION TYPE CANNOT release IF RELEASE NAME IS DEFINED
		if ( defined $versiontype and $versiontype eq "release" ) {
			$self->logWarning("versiontype cannot be 'release' if releasename is defined: $releasename");
			return;
		}

		#### VERSION TYPE MUST BE DEFINED IF RELEASE NAME IS DEFINED
		#### OK 		--versiontype major --releasename alpha		1.0.0 -> 2.0.0-alpha.1
		#### NOT OK		--releasename alpha				1.0.0  --XXXXX--> 1.0.0-alpha.1  !!!!!
		if ( not defined $versiontype and not $release ) {
			$self->logWarning("versiontype must be defined if releasename is defined");
			return;
		}
	
		#### MUST BE MAJOR, MINOR OR PATCH INCREMENT IF RELEASE NAME IS DEFINED AND NO RELEASE
		#### OK 		--versiontype major --releasename alpha	1.0.0 -> 2.0.0-alpha.1
		#### OK 		--versiontype build --releasename beta	1.0.0-alpha.1 ---> 1.0.0-beta.1+build1
		#### NOT OK 	--versiontype build --releasename alpha	1.0.0+build1 --XXX-> 1.0.0-alpha.1+build1
		#### NOT OK		--releasename alpha				1.0.0  --XXXXX--> 1.0.0-alpha.1  !!!!!
		my $isversion = 1;
		$isversion = 0 if
			defined $versiontype 
			and not $versiontype eq "major"
			and not $versiontype eq "minor"
			and not $versiontype eq "patch";
				
		if ( defined $versiontype
			and not $isversion
			and not $release
		) {
			$self->logWarning("versiontype must be major, minor or patch if releasename is defined");
			return;
		}

		if ( not $isversion
			and (
					( $releasename eq "alpha" and $release =~ /^(alpha|beta|rc)/ )
				or ( $releasename eq "beta" and $release =~ /^(beta|rc)/ )
				or ( $releasename eq "rc" and $release =~ /^rc/ )
			)
		) {
			$self->logWarning("releasename '$releasename' must be > current release: $release");
			return;
		}		

		if ( not defined $versiontype
			and (
					( $releasename eq "alpha" and $release =~ /^(alpha|beta|rc)/ )
				or ( $releasename eq "beta" and $release =~ /^(beta|rc)/ )
				or ( $releasename eq "rc" and $release =~ /^rc/ )
			)
		) {
			$self->logWarning("releasename '$releasename' must be > current release: $release");
			return;
		}		
	}

	#### INCREMENT VERSION IF VERSION TYPE IS DEFINED
	if ( defined $versiontype ) {
		if ( $versiontype eq "major" ) {
			$major++ ;
			$minor = 0;
			$patch = 0;
			$build = '';
			$release = '';
		}
		elsif ( $versiontype eq "minor" ) {
			$minor++;
			$patch = 0;
			$build = '';
			$release = '';
		}
		elsif ( $versiontype eq "patch" ) {
			$patch++;
			$build = '';
			$release = '';
		}
		elsif ( $versiontype eq "release") {
			if ( not defined $release or not $release ) {
				$self->logWarning("release must be defined if versiontype is release");
				return;
			}
			else {
				$release = $self->incrementMixed($release);
			}
		}
		elsif ( $versiontype eq "build" ) {
			if ( not $build ) {
				$build = "build.1";
			}
			else {
				#$self->logDebug("BEFORE build", $build);
				$build = $self->incrementMixed($build);
				$self->logDebug("AFTER incrementMixed, build", $build);
			}
		}
	}

	#### IF RELEASE NAME AND VERSION TYPE ARE DEFINED, JUST ATTACH THE
	#### RELEASE TO THE INCREMENTED VERSION TYPE
	####
	#### E.G., MAJOR VERSION WITH RELEASE NAME alpha:
	####
	#### 	0.9.2 -> 1.0.0-alpha
	####
	if ( defined $releasename ) {
		$release = $releasename;
		$release = $release . ".1" if $release !~ /\d+$/;
	}

	#### FINAL VERSION
	my $finalversion = "$major.$minor.$patch";
	$finalversion .= "-$release" if $release;
	$finalversion .= "+$build" if defined $versiontype and $versiontype eq "build";
	$self->logDebug("finalversion", $finalversion);

	return $finalversion;
}

method incrementMixed ($version) {
	my ($symbol, $number) = $version =~ /^(.*)(\d+)$/;
	$symbol = '' if not defined $symbol;
	$number = 0 if not defined $number;
	$number++;
	$version = $symbol . $number;

	return $version;
}


method parseSemVer ($version) {
	#$self->logDebug("version", $version);
	$version =~
		/^(\d+)\.(\d+)\.(\d+)(-alpha[\.\d]*|-beta[\.\d]*|-rc[\.\d]*)?(\+build[\.\d]*|\+[\.\d]*)?/;
	my $major 	= $1;
	my $minor 	= $2;
	my $patch 	= $3;
	my $release = $4;
	my $build  	= $5;

#$self->logDebug("major", $major);
#$self->logDebug("minor", $minor);
#$self->logDebug("patch", $patch);
#$self->logDebug("release", $release);
#$self->logDebug("build", $build);
	
	$release =~ s/^-// if defined $release;
	$build =~ s/^\+// if defined $build;
	$build =~ s/\.$// if defined $build;
	$release = '' if not defined $release;
	$build = '' if not defined $build;
	
	
	return ($major, $minor, $patch, $release, $build);
}

method getCommitCount ($repodir) {
	print "repodir is a file\n" and exit if -f $repodir;
	chdir($repodir) or die "Can't chdir to repodir: $repodir\n";
	my $iteration = `git log --pretty=format:'' | wc -l`;
	$iteration =~ s/\s+//g;
	$iteration = "0" x ( 5 - length($iteration) ) . $iteration;
	
	return $iteration;	
}

method getBuildId ($repodir) {
	#### 2. GET THE SHORT SHA KEY AS THE BUILD ID
	my $command = "git rev-parse --short HEAD";
	my $buildid = $self->repoCommand($command);
	$buildid =~ s/\s+//g;
	print "buildid: $buildid\n";
	
	return $buildid;
}

method createVersionDir ($repodir, $package, $version) {
#### CREATE THE PACKAGE DIR AND VERSION SUBDIR
	my $versiondir = "$repodir/$package/$version";
	`mkdir -p $versiondir`;
	print "Can't create versiondir: $versiondir\n"
    and exit if not -d $versiondir;
}




1;